Add dropdown version of Filter by Rating

kmanijak opened the pull request:
**Main changes:**
- add dropdown version to the Filter by Rating
- add option for `multiple` or `single` choice to align with Filter by Attribute

**Minor changes:**
- alignment between Filter by Attribute and Rating:
  - Editor settings options: wording and settings order
  - selected options have the same styling
- Dropdown has rounder border, previously there was little gaps on the corners

Important notice: to be able to provide to the dropdown (`FormTokenField` component from Gutenberg) the options more complex than a string, a bit hacky solution has been introduced. `displayTransform` is a place where string value is transformed into what is presented to the user. In this PR JSX element is returned ([check here](https://github.com/woocommerce/woocommerce-blocks/pull/7771/files#diff-c253e7e8181d1c499245ce3afcb4674b71039ff16e1c5b6cf07ffbc141dc2bc0R385)). That solution may stop to work in case of Gutenberg update.

Also, seems like this may have a consequence on the built-int accessibility mechanism. There's `messages` prop passed to the `FormTokenField`, which includes messages that are read in case thee filter is added/removed. That seems to not work in here, which may be an obstacle to go with that solution.

Fixes #7769 #7882 

#### Accessibility

<!-- If you've changed or added any interactions, check off the appropriate items below. You can delete any that don't apply. Use this space to elaborate on anything if needed. -->

- [x] I've tested using only a keyboard (no mouse)
- [x] I've tested using a screen reader

Known problem:
- Screen Reader doesn't read when filter is added/removed.

### Screenshots
Watch the video of Dropdown component usage:

https://user-images.githubusercontent.com/20098064/205309205-b0b69334-541a-42bd-881e-455db517b70e.mov

| Case | Before | After |
| ------ | ----- | ----- |
|    Filter by Rating Editor settings    |   <img width="283" alt="Screen Shot 2022-12-02 at 14 39 20 PM" src="https://user-images.githubusercontent.com/20098064/205305555-0d988c66-6a92-432a-9ca6-61d2b62c492c.png">    |    <img width="278" alt="Screen Shot 2022-12-02 at 14 36 25 PM" src="https://user-images.githubusercontent.com/20098064/205304976-f25d2e0a-dd32-43bf-8762-5a901c1393ea.png">  |
| Dropdown borders |   <img width="766" alt="Screen Shot 2022-12-02 at 14 39 05 PM" src="https://user-images.githubusercontent.com/20098064/205305527-36bd5789-fe7a-4ed1-a9ce-3ceecbf09ca1.png">   | <img width="742" alt="Screen Shot 2022-12-02 at 14 37 22 PM" src="https://user-images.githubusercontent.com/20098064/205305181-eaa1f182-5ab2-48dc-9ef6-2ef3645aa06c.png"> |  

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [x] E2E tests - single one

#### User Facing Testing

Check out [the video above](https://user-images.githubusercontent.com/20098064/205309205-b0b69334-541a-42bd-881e-455db517b70e.mov) which will help you go through the testing instructions

Prerequisites:
1. Have at least few products with some ratings
2. Go to Editor
3. Add Filters pattern
4. Add All Products

##### Dropdown component - single
1. Change settings of Filter by Rating to: single / dropdown
2. Open dropdown
  - **Expected**: dropdown has rating options to choose from
3. Choose one option
  - **Expected**: page refreshes, filter is applied: visible in Active Filters and Products are filtered out
4. Choose other option
  - **Expected**: previous filter is replaced by the new one, page refreshes, filter is applied: visible in Active Filters and Products are filtered out

##### Dropdown component - multiple
1. Change settings of Filter by Rating to: multiple / dropdown
2. Open dropdown
  - **Expected**: dropdown has rating options to choose from
3. Choose one option
  - **Expected**: page refreshes, filter is applied: visible in Active Filters and Products are filtered out
4. Choose other option
  - **Expected**: new filter is added to the new one, page refreshes, both filters ("or") is applied: visible in Active Filters and Products are filtered out

##### List component
1. Change settings of Filter by Rating to: list
2. Repeat the above for both: single and multiple

##### Apply button
1. In the Editor choose multiple Dropdown again and enable "Show 'Apply filters' button" option
2. In the frontend choose rating filters
  - **Expected**: filters are not applied until "Apply" button is clicked. Once clicked page refreshes and filters are applied

##### Other scenarios
Try to break it! 😉 

### WooCommerce Visibility

* [ ] WooCommerce Core
* [x] Feature plugin
* [ ] Experimental

### Changelog

> Filter by Rating: Add dropdown version.
> Filter by Rating: Add option to choose multiple or single option.
> Filters: Fix the little gaps on the border corners in the filters Dropdown version.

kmanijak committed:
Rename the setting section to match Filter by Attribute

Rename section from  to

kmanijak committed:
Add Display Style toggle to the block settings

kmanijak committed:
Set  as a default value of displayStyle for Filter by Rating

kmanijak committed:
Extend parsing function with displayStyle property

kmanijak committed:
Add dropdown variant of Filter by Rating - not yet working, but compiling

kmanijak committed:
Extend the Editor settings with selectType: single or multiple and queryType: or or and

kmanijak committed:
Adjust the styles of a FormTokenField to other dropdown

kmanijak committed:
Align Editor settings order and wording between Filter by Attribute and Rating

kmanijak committed:
Align styles of the dropdowns between Filter by Attribute and Rating

Additionally, fix the issue with cut off corners of dropdown borders

kmanijak committed:
Remove unused property from options in Filter by Attribute

kmanijak committed:
Provide translated screen reader messages and placeholder

kmanijak committed:
Add logic to handle clicking the options in dropdown

kmanijak committed:
Revert the label property, which is necessary to display checkbox list

kmanijak committed:
Make classname for Rating component optional and remove styling in Filter by Rating

kmanijak committed:
Cover the case when filter with zero products needs string methods for comparison

kmanijak committed:
Handle Typescript errors and add ts-ignore annotations when necessary

kmanijak committed:
Remove option to choose query type as it doesn't make sense in the context of rating

kmanijak committed:
Remove explicit key prop assignment in Rating component

kmanijak committed:
Remove leftover property in type definition

kmanijak committed:
Refactor JSX element extension with custom properties to more robust way with object copy

kmanijak committed:
Filter by rating tests (#7845)

* Filter by Rating tests:
  - List single
  - List multiple
  - Dropdown single
  - Dropdown multiple

kmanijak committed:
Merge branch 'trunk' into add/filter-by-rating-dropdown

kmanijak commented:
@Aljullu , @danieldudzic , PR is ready to be reviewed, tests are included 🙌

kmanijak committed:
Adjust the font-size of Filter by Rating dropdown to the rest of the filters

kmanijak committed:
Fix E2E tests after the Editor settings label has been changed

Aljullu commented on assets/js/blocks/attribute-filter/style.scss:
```diff
@@ -76,12 +76,17 @@
+
+		.components-form-token-field {
+			border-radius: inherit;
+		}
```
Good catch!

Aljullu commented on assets/js/blocks/rating-filter/block.tsx:
```diff
@@ -204,13 +255,6 @@ const RatingFilterBlock = ( {
 				return {
 					label: (
 						<Rating
-							className={
-								Array.from( productRatings ).includes(
-									item?.rating?.toString()
-								)
-									? 'is-active'
-									: ''
-							}
```
Is it intentional that you removed the `is-active` class? I think it's used in the list display style to highlight the selected options.

kmanijak replied:
> it's used in the list display style to highlight the selected options.

That's true, but that was the only filter that did that. I wanted to align filters UI, so I consulted with @vivialice and we agreed to remove the highlight in Filter by Rating.

Aljullu replied:
Should we remove the associated styling, then?

https://github.com/woocommerce/woocommerce-blocks/blob/8c857e2422d68479e7f5d798184fa19931e1eb07/assets/js/blocks/rating-filter/style.scss#L14-L19

Aljullu commented on assets/js/blocks/rating-filter/block.tsx:
```diff
@@ -293,20 +331,146 @@ const RatingFilterBlock = ( {
+								.map( ( option ) => option.value ) }
+							disabled={ isLoading }
+							placeholder={ __(
+								'Select Rating',
```
I don't have a strong opinion. But if we follow [these rules](https://github.com/WordPress/gutenberg/blob/086b77ed409a70a6c6a6e74dee704851eff812f2/docs/contributors/documentation/copy-guide.md#five-pay-attention-to-capitalization), I think we should use sentence capitalization, here.
```suggestion
								'Select rating',
```

kmanijak replied:
I took the convention from Filter by Attribute, but I agree with you. I changed both [in this commit](https://github.com/woocommerce/woocommerce-blocks/pull/7771/commits/6a09198ccc997eb68fd47c5f4445f88ad5ca916d).

Aljullu replied:
Hmmm, thinking about it twice, maybe I would undo this last commit (sorry for the back and forth :sweat: ). In the case of attributes, we can't know if the attribute name needs to be capitalized (ie: if the attribute is `CD style` we wouldn't want to lowercase it to `Select cd style`). 

And then, if we leave the attribute name capitalized, I see that it makes sense to keep `Select Rating` consistent.

Aljullu commented on assets/js/blocks/rating-filter/utils.ts:
```diff
@@ -18,10 +23,28 @@ export const getActiveFilters = ( queryParamKey = 'filter_rating' ) => {
 	return parsedParams;
 };
 
+export function generateUniqueId() {
+	return Math.floor( Math.random() * Date.now() );
+}
+
+export const formatSlug = ( slug: string ) =>
+	slug
+		.trim()
+		.replace( /\s/g, '-' )
+		.replace( /_/g, '-' )
+		.replace( /-+/g, '-' )
+		.replace( /[^a-zA-Z0-9-]/g, '' );
```
Given that _Filter by Stock_ will also need these utils. I wonder if we should extract them to a common file. What do you think?

kmanijak replied:
I agree. But I would postpone it until both blocks are merged. I think syncing two PRs may be time-consuming, while it will take a moment to extract it when both are merged. What do you think?

Aljullu replied:
Sounds good! :100:

Aljullu commented on assets/js/blocks/rating-filter/test/block.tsx:
```diff
@@ -0,0 +1,497 @@
+		selectType: 'multiple',
+	} );
+
+describe( 'RatingFilterBlock', () => {
```
I would try to use the 'official' block name here:
```suggestion
describe( 'Filter by Rating block', () => {
```
(This probably needs to be changed in the Filter by Attribute block tests as well)

kmanijak replied:
Sure thing! Changed in both: Filter by Rating and Attribute [in this commit](https://github.com/woocommerce/woocommerce-blocks/pull/7771/commits/df799fd4c89eb17271745d649ddafec0413a4740).

Aljullu commented on assets/js/blocks/rating-filter/test/block.tsx:
```diff
@@ -0,0 +1,497 @@
+
+			const dropdown = getDropdown();
+
+			if ( dropdown ) {
```
Is there any case where this `if` might be false?

kmanijak replied:
We expect it to always be truthy, but Typescript complained, so I used `if` statement.

I'm not sure now, maybe I shouldn't wrap it just for Typescript 🤔 I don't have a strong opinion.

Aljullu replied:
Aaah, makes sense. Let's leave it like this, then.

Aljullu commented on assets/js/blocks/rating-filter/test/block.tsx:
```diff
@@ -0,0 +1,497 @@
+
+			const rating4Suggestion = getRating4Suggestion();
+
+			if ( rating4Suggestion ) {
```
Same question here. :slightly_smiling_face:

Aljullu reviewed (commented):
Thanks for working on this, @kmanijak! That's a big feature and maybe I would have split the minor changes into separate PRs so they could be merged without waiting for the other changes to be approved. Not a big deal, but mentioning it for future PRs; IMO, the smaller the PRs are, the better.

I really appreciate that you included improvements to existing code and to the Filter by Attribute block. Also, good job with the tests! :clap: 

I tested it and overall it's working well, but I noticed some small issues, I added a video for each one of them:

* [Dropdown with single option: the input field might overflow the wrapper](https://user-images.githubusercontent.com/3616980/206178258-45227687-5afe-4205-b73a-109471a27a1a.webm).
* [Dropdown with multiple options when all options have been added](https://user-images.githubusercontent.com/3616980/206177760-73419348-0f61-4fd6-a52a-5700a5cc140a.webm): I think we should hide the input; otherwise it feels like you can still add more options, but clicking on the control doesn't do anything.
* This is a nitpick, but [in the dropdown the options are sorted in a descending order, but once they are selected, they are ordered in a ascending order](https://user-images.githubusercontent.com/3616980/206178956-650002b1-a45c-40c2-aab9-909d17a679a7.webm). (Ie: in the dropdown 5 stars is rendered above 4 stars but in the input field 4 starts is rendered before 5 stars)

IMO the first one is blocking and the second and third ones can be worked on in follow-up PRs. But I'm happy if you work on them as part of this PR, if you prefer.

kmanijak committed:
Improve sorting in Filter by Rating dropdown

With this change the order of options is always descending: in the chips, suggestions, active filters and URL

kmanijak committed:
Use sentence capitalization in dropdown placeholders

Apply capitzalization according to the Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/086b77ed409a70a6c6a6e74dee704851eff812f2/docs/contributors/documentation/copy-guide.md\#five-pay-attention-to-capitalization

kmanijak committed:
Change the name of the filter blocks in the test files

kmanijak committed:
Prevent single dropdown input to overflow the wrapper

The issue is fixed in three filters, hence a new shared styles file is created at the path assets/js/blocks/shared/styles/style.css

kmanijak commented:
@Aljullu , thanks for thorough review! 🙏

### 1. Dropdown with single option: the input field might overflow the wrapper
> [Dropdown with single option: the input field might overflow the wrapper](https://user-images.githubusercontent.com/3616980/206178258-45227687-5afe-4205-b73a-109471a27a1a.webm).

It's fixed [in this commit](https://github.com/woocommerce/woocommerce-blocks/pull/7771/commits/f1d4401d864470809b7ac2dcd9f610609c2eddd8). I didn't want to repeat that fix for each dropdown, so I created a new shared style file on the `block` directory level (path: `assets/js/blocks/shared/styles/style.css`). The styles are applied to Filter by Attribute, Rating and Stock (even though it's not yet merged).
Luckily that fixed another issue I've submitted earlier: https://github.com/woocommerce/woocommerce-blocks/issues/7882

[Here's a video with updated block](https://user-images.githubusercontent.com/20098064/206442724-5c6be8e2-239d-45bb-a38a-488d1959e5bd.mov)

### 2. Dropdown with multiple options when all options have been added

> [Dropdown with multiple options when all options have been added](https://user-images.githubusercontent.com/3616980/206177760-73419348-0f61-4fd6-a52a-5700a5cc140a.webm): I think we should hide the input; otherwise it feels like you can still add more options, but clicking on the control doesn't do anything.

I think we can tackle it separately to bloat the PR no more. I created an issue: https://github.com/woocommerce/woocommerce-blocks/issues/7889

### 3. In the dropdown the options are sorted in a descending order, but once they are selected, they are ordered in a ascending order

> This is a nitpick, but [in the dropdown the options are sorted in a descending order, but once they are selected, they are ordered in a ascending order](https://user-images.githubusercontent.com/3616980/206178956-650002b1-a45c-40c2-aab9-909d17a679a7.webm). (Ie: in the dropdown 5 stars is rendered above 4 stars but in the input field 4 starts is rendered before 5 stars)

It's fixed in [this commit](https://github.com/woocommerce/woocommerce-blocks/pull/7771/commits/1ea0678754f0d8831f361bd18fbac6d035c1257c). Order is always descending in all places: active filters, dropdown options, chosen options and in the URL.

[Here's a video with updated block](https://user-images.githubusercontent.com/20098064/206442472-648c664c-9aae-4c4c-bc74-d50924f6df3f.mov)

Aljullu commented on assets/js/blocks/shared/styles/style.scss:
```diff
@@ -0,0 +1,15 @@
+			}
+		}
+	}
+}
```
Not needed for this PR, but I wonder if at some point we should create a `filters` directory under `assets/js/blocks/`. That would make it easier to share styles and logic between them, like this one. Similar to what was done with Featured Product and Featured Category, which have a common folder (`featured-items`).

kmanijak replied:
I created a `assets/js/blocks/shared/styles/style.css` file (and added a mixin dedicated to filters), but in the near future we should make a dedicated path (`assets/js/blocks/filters/*`) with shared JS and SCSS code. There's too much code overlap between three filters to not combine it 😆  The issue is already created: https://github.com/woocommerce/woocommerce-blocks/issues/7875 and that's definitely the direction we should follow.

Aljullu reviewed (approved):
Works like a charm, thanks @kmanijak!

I'm pre-approving, but I left a couple of comments that I think should be adressed before merging:

* https://github.com/woocommerce/woocommerce-blocks/pull/7771#discussion_r1044369179
* https://github.com/woocommerce/woocommerce-blocks/pull/7771#discussion_r1044377583

kmanijak committed:
Remove the unused styles of active options in Filter by Rating

Chosen options in Filter by Rating list have been bolded. That was the only filter that was doing it. Class was removed, but not the styles themselves which became obsolete. Removing the leftover styles then

kmanijak committed:
Revert the filters placeholder capitalization. Keep the kebab case instead of sentence capitalization

Not knowing what may be attribute, there may be a case first capital letter will be expected

kmanijak committed:
Change the approach of shared styles to fix the dropdown issue in filters

Instead of using direct classes, rewrite it to mixin and include in the filters

kmanijak commented:
Thank you for a great review @Aljullu! 🎉 

> I'm pre-approving, but I left a couple of comments that I think should be adressed before merging:

Great, both comments addressed.


PS. Good catch with the capitalization!

> ie: if the attribute is "CD style" we wouldn't want to lowercase it to Select "cd style"

kmanijak committed:
Merge branch 'trunk' into add/filter-by-rating-dropdown

kmanijak commented:
Thank you for a great review @Aljullu! 🎉 

> I'm pre-approving, but I left a couple of comments that I think should be adressed before merging:

Great, both comments addressed.


PS. Good catch with the capitalization!

> ie: if the attribute is "CD style" we wouldn't want to lowercase it to Select "cd style"

kmanijak committed:
Merge branch 'trunk' into add/filter-by-rating-dropdown

kmanijak committed:
Merge branch 'trunk' into add/filter-by-rating-dropdown

gigitux committed:
Merge branch 'trunk' into add/filter-by-rating-dropdown

kmanijak committed:
Fix the Filter by Rating with Products block

The problem was with the Products block, that there was incorrect initial value of product ratings query

kmanijak committed:
Align the logic of getting the filters from URL for Filter by Rating with Filter by Stock

gigitux reviewed (approved):
I tested only the feature and I didn't review the code.

I didn't find any issue! Nice work! LGTM!

kmanijak committed:
Merge branch 'trunk' into add/filter-by-rating-dropdown
